File: sdks/ruby/lib/novabilling/types/invoice_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class InvoiceResponse < Internal::Types::Model
      field :id, -> { String }, optional: false, nullable: false
      field :invoice_number, -> { String }, optional: false, nullable: false, api_name: "invoiceNumber"
      field :subscription_id, -> { String }, optional: true, nullable: false, api_name: "subscriptionId"
      field :customer_id, -> { String }, optional: false, nullable: false, api_name: "customerId"
      field :amount, -> { String }, optional: false, nullable: false
      field :currency, -> { String }, optional: false, nullable: false
      field :status, -> { Novabilling::Types::InvoiceResponseStatus }, optional: false, nullable: false
      field :due_date, -> { String }, optional: false, nullable: false, api_name: "dueDate"
      field :paid_at, -> { String }, optional: true, nullable: false, api_name: "paidAt"
      field :pdf_url, -> { String }, optional: true, nullable: false, api_name: "pdfUrl"
      field :metadata, -> { Internal::Types::Hash[String, Object] }, optional: true, nullable: false
      field :customer, -> { Novabilling::Types::InvoiceCustomerResponse }, optional: true, nullable: false
      field :created_at, -> { String }, optional: false, nullable: false, api_name: "createdAt"
      field :updated_at, -> { String }, optional: false, nullable: false, api_name: "updatedAt"
    end
  end
end

File: sdks/csharp/src/NovaBilling/Types/InvoiceResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record InvoiceResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("invoiceNumber")]
    public required string InvoiceNumber { get; set; }

    [JsonPropertyName("subscriptionId")]
    public string? SubscriptionId { get; set; }

    [JsonPropertyName("customerId")]
    public required string CustomerId { get; set; }

    /// <summary>
    /// Decimal amount as string
    /// </summary>
    [JsonPropertyName("amount")]
    public required string Amount { get; set; }

    [JsonPropertyName("currency")]
    public required string Currency { get; set; }

    [JsonPropertyName("status")]
    public required InvoiceResponseStatus Status { get; set; }

    [JsonPropertyName("dueDate")]
    public required string DueDate { get; set; }

    [JsonPropertyName("paidAt")]
    public string? PaidAt { get; set; }

    [JsonPropertyName("pdfUrl")]
    public string? PdfUrl { get; set; }

    /// <summary>
    /// Line items, plan info, discounts
    /// </summary>
    [JsonPropertyName("metadata")]
    public Dictionary<string, object?>? Metadata { get; set; }

    [JsonPropertyName("customer")]
    public InvoiceCustomerResponse? Customer { get; set; }

    [JsonPropertyName("createdAt")]
    public required string CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public required string UpdatedAt { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
